### PR TITLE
 Add comprehensive test suite to cover core extraction, scoring, and APIs  #11

### LINKED ISSUE #11: [TESTING] No unit or integration tests — add pytest suite with mock EML fixtures
## Summary
The repository has no `tests/` directory and no test runner config. Regressions in regex patterns, defanging logic, or risk-scoring are undetectable without running the full tool against a live email.

## Minimum coverage needed

| Module | Tests |
|---|---|
| `extract_iocs()` | URL, domain, IPv4/IPv6 extraction; private IP filtering |
| `defang_url/domain/ip()` | Correct defanging for each type |
| `_extract_auth_result()` | SPF pass/fail/softfail from realistic headers |
| `calculate_risk()` | All four risk tiers |
| `extract_attachments()` | SHA-256 accuracy, filename sanitisation |
| `_virustotal_get()` | Mocked 200 / 401 / 429 / 404 |
| `query_abuseipdb()` | Score >= 75 → malicious |
| `safe_md()` | Pipe and newline injection |

## Structure
```
pip install pytest pytest-mock responses
tests/
  conftest.py
  test_ioc.py
  test_defang.py
  test_headers.py
  test_scoring.py
  test_attachments.py
  test_api_clients.py
  test_report.py
```

### PR DESCRIPTION
**Pull request by Deadpool2000**
#11 While looking at your codebase, I realized that we didn't have a foundational test suite set up for phish_extractor. Without it, any future tweaks to our regex patterns, defanging logic, or risk scoring could easily introduce silent regressions unless we manually ran the script against a live .eml file every time.

To fix this and make the project easier to maintain, I’ve introduced pytest and built out our baseline test coverage.

Here is exactly what I did:

- Project Structure: I set up a tests/ directory with a standard modular structure (including an empty conftest.py for future shared fixtures).

- IOC & Defanging: I wrote tests to ensure extract_iocs() correctly parses URLs, domains, and both IPv4/IPv6 addresses, while properly filtering out private/internal IPs. I also verified that our defang_* functions correctly neutralize links and IPs.

- Header & Attachment Parsing: Added coverage for  _extract_auth_result()  to ensure we accurately flag SPF/DKIM/DMARC passes and fails. I also tested extract_attachments() to verify it reliably grabs filenames and computes accurate SHA-256 hashes without needing to write files to disk.

- Mocked Threat Intel APIs: I brought in the responses library so I could safely mock our external calls to VirusTotal and AbuseIPDB. I made sure we're correctly handling 200 OKs, authentication errors (401), rate limits (429), and not-found (404) scenarios, and validated that our >= 75 confidence threshold for AbuseIPDB works as intended.

- Risk Scoring & Reporting: Finally, I added tests for our calculate_risk()  tiers to make sure all four levels (LOW to CRITICAL) map correctly, and checked safe_md() to prevent any odd table-breaking markdown injection in the final report.

Everything is currently passing locally via 'PYTHONPATH=. pytest tests/'. 

Let me know if you'd like me to add any additional edge cases!

### CODE REVIEW

**Review by gurvinny — COMMENTED** [review]
test

**Review by gurvinny — CHANGES_REQUESTED** [review]
## Code Review

Thanks for putting this together, @Deadpool2000 — this is a well-structured contribution that covers all major modules and uses correct mocking patterns. The tests accurately reflect the source implementation. A few things to address before merge:

---

### Bugs / Correctness

**1. Misleading comment in `test_defang_ip`**

```python
assert defang_ip("2001:db8::1") == "2001:db8::1"  # IPv6 has no dots, should remain unchanged if no dots
```

The comment is misleading — the very next assertion shows IPv6 *can* have dots (`::ffff:192.168.1.1`). Please remove or clarify the comment.

**2. Fragile inline score comment in `test_calculate_risk_medium`**

```python
# score = 2 (spf fail) + 1 (error) = 3 -> MEDIUM
```

If scoring weights change, this comment silently becomes wrong. Consider removing it or only documenting the threshold boundary, not the raw point values.

---

### Missing Tests

**3. No AbuseIPDB error-path tests**

`test_api_clients.py` tests VirusTotal error codes (401, 429, 404) but provides zero coverage for AbuseIPDB error handling, which has the same paths in the source (`phish_extractor.py:727-735`). Please add equivalent tests for at least 401 and 429.

**4. No scoring boundary value tests**

The scoring thresholds are `<=2 LOW | 3-5 MEDIUM | 6-9 HIGH | 10+ CRITICAL`. There are no tests at the exact boundary values (score=2, 3, 5, 6, 9, 10). Boundary values are the most common source of off-by-one bugs.

**5. `softfail`/`none` DMARC result not tested**

The source only adds +2 for `== "fail"`. A `dmarc=softfail` result contributes 0 points. Worth a test to confirm this is intentional — or a question to the maintainer about whether `softfail` should score.

**6. No VT timeout / connection error tests**

The source handles `requests.exceptions.Timeout` and `RequestException` but neither path is covered. You can trigger these with:
```python
responses.add(responses.GET, endpoint, body=requests.exceptions.Timeout())
```

---

### Code Quality

**7. Testing a private function directly (`test_headers.py`)**

```python
from phish_extractor import _extract_auth_result
```

`_extract_auth_result` is a private implementation detail (underscore prefix). Testing it directly couples tests to internals — a rename or refactor breaks this test even if public behaviour is unchanged. Prefer testing through the public `extract_headers()` function instead.

**8. Repeated URL-ID setup in every VT test**

All four VT tests in `test_api_clients.py` duplicate this block:

```python
url = "http://evil.com"
url_id = base64.urlsafe_b64encode(url.encode()).decode().rstrip("=")
endpoint = f"https://www.virustotal.com/api/v3/urls/{url_id}"
```

This belongs in a shared pytest fixture in `conftest.py`.

**9. `conftest.py` is empty**

The file contains only `# conftest.py`. Shared fixtures (sample `EmailMessage`, API key monkeypatches, common `responses` setup) should live here to reduce duplication across test files.

---

### Missing Infrastructure

**10. No `pytest.ini` / `pyproject.toml` test config**

No pytest configuration declares the test directory, markers, or coverage options. A minimal `[tool.pytest.ini_options]` block would make the suite reproducible for all contributors.

**11. No test dependency declaration**

The tests require `pytest` and `responses` but neither is declared in a `requirements-test.txt` or `extras_require`. A new contributor will not know what to install.

---

### Summary

| Category | Count |
|---|---|
| Bugs / Correctness | 2 |
| Missing tests | 4 |
| Code quality | 3 |
| Missing infrastructure | 2 |

The core test logic is accurate and well thought-out — once the gaps above are addressed this will be a great addition to the project.

File: tests/test_api_clients.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    
+    responses.add(responses.GET, endpoint, status=404)
+    
+    result = query_virustotal_url(url)
+    assert result.error == "Not found in VirusTotal database"
+
+@responses.activate
+def test_query_abuseipdb_malicious(monkeypatch):
+    monkeypatch.setattr('phish_extractor.ABUSEIPDB_API_KEY', 'dummy_key')
+
```
Comment (gurvinny) [inline tests/test_api_clients.py]: Good boundary test — `>= 75` is the threshold in the source. Consider also adding error-path tests for AbuseIPDB, mirroring what you did for VT. The source handles `401` and `429` for AbuseIPDB (`phish_extractor.py:727-735`) but neither path has any test coverage yet.

File: tests/test_attachments.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    assert attachments[0].size_bytes == len(payload_bytes)
+    assert attachments[0].sha256 == hashlib.sha256(payload_bytes).hexdigest()
+
+def test_extract_attachments_no_filename():
+    msg = EmailMessage()
+    msg['Subject'] = 'Test with nameless attachment'
+    
+    msg.set_content('Please see the attached file')
+    payload_bytes = b'test'
+    msg.add_attachment(payload_bytes, maintype='application', subtype='octet-stream')
```
Comment (gurvinny) [inline tests/test_attachments.py]: The source also picks up `inline` dispositions (`phish_extractor.py:499`). Worth adding a test where the attachment is added with `Content-Disposition: inline` to confirm it is extracted correctly.

File: tests/test_defang.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    assert defang_url("https://malicious.co.uk/x.php") == "hxxps://malicious[.]co[.]uk/x.php"
+    assert defang_url("https://1.2.3.4/abc") == "hxxps://1[.]2[.]3[.]4/abc"
+
+def test_defang_domain():
+    assert defang_domain("evil.com") == "evil[.]com"
+    assert defang_domain("malicious.co.uk") == "malicious[.]co[.]uk"
+
+def test_defang_ip():
+    assert defang_ip("8.8.8.8") == "8[.]8[.]8[.]8"
+    assert defang_ip("2001:db8::1") == "2001:db8::1" # IPv6 has no dots, should remain unchanged if no dots
```
Comment (gurvinny) [inline tests/test_defang.py]: The trailing comment is misleading — it says "no dots" but the very next assertion (`::ffff:192.168.1.1`) shows IPv6 *can* have dots. Either remove the comment or reword it:

```python
assert defang_ip("2001:db8::1") == "2001:db8::1"  # pure hex IPv6: no dots to replace
```

**Review by gurvinny — COMMENTED** [review]
Inline comments added per file — see individual line notes below.

**Comment by Deadpool2000** [comment]
1. test_defang_ip

I updated the comment in tests/test_defang.py to clarify that an IPv6 address without dots stays unchanged, acknowledging the test directly below it which checks that mapped IPv4 (with dots) is successfully defanged.

2. Removed fragile scoring from test_score_medium: 

I replaced the hardcoded point calculation logic with # Reaches MEDIUM risk threshold in tests/test_scoring.py For consistency, I proactively made the same change for the HIGH and CRITICAL risk tests in that file, ensuring that no score values are hardcoded in test comments moving forward.

**Comment by gurvinny** [comment]
Thanks for putting this together, and apologies for the slow reply.

Closing as superseded: a unittest-based suite landed in July covering extraction, headers, IOC handling, scoring and reporting, which overlaps most of what is here. This branch also imports `responses`, which is not declared in requirements.txt and not installed by CI, so it would not run as-is.

Appreciate the effort.
